stan/math/prim/fun/beta.hpp

```diff
@@ -65,7 +65,8 @@ inline return_type_t<T1, T2> beta(const T1 a, const T2 b) {
  * @param b Second input
  * @return Beta function applied to the two inputs.
  */
-template <typename T1, typename T2, require_any_container_t<T1, T2>* = nullptr>
+template <typename T1, typename T2, require_any_container_t<T1, T2>* = nullptr,
+          require_all_not_var_matrix_t<T1, T2>* = nullptr>
```

Comment:
> So in the case of beta(varmat, matvar), the matvar gets promoted to a varmat and the rev specialisation is called? Is it worth testing whether this overhead is slower than just looping over the varmat?

So in the case of any `f(A, B)` where A or B is a `var<matrix>` and the return is a matrix type the return will be a `var<matrix>`. The input `matrix<var>` itself is not promoted though. If you look at the beta that accepts at least one `var<Matrix>` you'll see the actual type of the input matrix is preserved, then the output is made into a `var<Matrix>`. I can bench this if you like but it should always be faster, especially for ops like 

```
    return make_callback_var(
        beta(arena_a.val(), arena_b.val()),
        [arena_a, arena_b, digamma_ab](auto& vi) mutable {
          const auto adj_val = (vi.adj().array() * vi.val().array()).eval();
          arena_a.adj().array()
              += adj_val * (digamma(arena_a.val().array()) - digamma_ab);
          arena_b.adj().array()
              += adj_val * (digamma(arena_b.val().array()) - digamma_ab);
        });
```

Where since the return is a `var<Matrix>` we get nicely lined up memory for calculating `adj_val`, then for either `a` or `b` being a `var<Matrix>` we get the same thing. For the one that is a `Matrix<var>` it just does the normal thing